source/mir/random/tinflex/internal/area.d

```diff
@@ -0,0 +1,669 @@
+module mir.random.tinflex.internal.area;
+
+import mir.random.tinflex.internal.types : Interval;
+import std.traits : ReturnType;
+
+/**
+Determines the hat and squeeze function of an interval.
+Based on Theorem 1
+*/
+void determineSqueezeAndHat(S)(ref Interval!S iv)
+in
+{
+    assert(iv.lx < iv.rx, "invalid interval");
+}
+body
+{
+    import mir.utility.linearfun : linearFun, secant, LinearFun;
+    import mir.random.tinflex.internal.types : determineType, FunType;
+
+    enum sec = "secant(iv.lx, iv.rx, iv.ltx, iv.rtx)";
+    enum t_l = "linearFun!S(iv.lt1x, iv.lx, iv.ltx)";
+    enum t_r = "linearFun!S(iv.rt1x, iv.rx, iv.rtx)";
+
+    // could potentially be saved for subsequent calls
+    FunType type = determineType(iv);
+    with(FunType) with(iv)
+    switch(type)
+    {
+        case T1a:
+            squeeze = mixin(t_r);
+            hat = mixin(t_l);
+            break;
+        case T1b:
+            squeeze = mixin(t_l);
+            hat = mixin(t_r);
+            break;
+        case T2a:
+            squeeze = mixin(sec);
+            hat = mixin(t_l);
+            break;
+        case T2b:
+            squeeze = mixin(sec);
+            hat = mixin(t_r);
+            break;
+        case T3a:
+            squeeze = mixin(t_r);
+            hat = mixin(sec);
+            break;
+        case T3b:
+            squeeze = mixin(t_l);
+            hat = mixin(sec);
+            break;
+        case T4a:
+            if (iv.lx == -S.infinity)
+            {
+                squeeze = squeeze.init;
+                hat = mixin(t_r);
+                break;
+            }
+            if (iv.rx == +S.infinity)
+            {
+                squeeze = squeeze.init;
+                hat = mixin(t_l);
+                break;
+            }
+            squeeze = mixin(sec);
+            hat = iv.ltx > iv.rtx ? mixin(t_l) : mixin(t_r);
+            break;
+        case T4b:
+            squeeze = iv.ltx < iv.rtx ? mixin(t_l) : mixin(t_r);
```

Comment:
current code:

![image](https://cloud.githubusercontent.com/assets/4370550/16506110/8e715388-3f20-11e6-94f2-62cb0288e9b8.png)


after changing it according to the paper:

![image](https://cloud.githubusercontent.com/assets/4370550/16506099/85c01bde-3f20-11e6-9043-421835101b98.png)